src/Stratis.Bitcoin/Connection/ConnectionManagerController.cs

```diff
@@ -20,65 +21,98 @@ public class ConnectionManagerController : FeatureController
         /// <summary>Instance logger.</summary>
         private readonly ILogger logger;
 
-        public ConnectionManagerController(IConnectionManager connectionManager,
-            ILoggerFactory loggerFactory) : base(connectionManager: connectionManager)
+        /// <summary>Network peer banning functionality.</summary>
+        private readonly IPeerBanning peerBanning;
+
+        public ConnectionManagerController(
+            IConnectionManager connectionManager,
+            IPeerBanning peerBanning,
+            ILoggerFactory loggerFactory)
+            : base(connectionManager: connectionManager)
         {
             Guard.NotNull(this.ConnectionManager, nameof(this.ConnectionManager));
+
             this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.peerBanning = peerBanning;
         }
 
         /// <summary>
         /// RPC method for adding a node connection.
         /// </summary>
-        /// <param name="command">The command to run. {add, remove, onetry}</param>
         /// <param name="endpointStr">The endpoint in string format.</param>
+        /// <param name="command">The command to run. {add, remove, onetry}</param>
         /// <returns><c>true</c> if successful.</returns>
         /// <exception cref="ArgumentException">Thrown if unsupported command given.</exception>
         [ActionName("addnode")]
         [ApiExplorerSettings(IgnoreApi = true)]
         [ActionDescription("Adds a node to the connection manager.")]
-        public bool AddNodeRPC(string endpointStr, string command)
+        public async Task<AddNodeRpcResult> AddNodeRPCAsync(string endpointStr, string command)
         {
+            var result = new AddNodeRpcResult();
+
             IPEndPoint endpoint = endpointStr.ToIPEndPoint(this.ConnectionManager.Network.DefaultPort);
             switch (command)
             {
                 case "add":
-                    this.ConnectionManager.AddNodeAddress(endpoint);
-                    break;
+                    {
+                        if (this.peerBanning.IsBanned(endpoint))
+                            result.ErrorMessage = $"{endpoint} is banned.";
```

Comment:
Use outer `try/catch` and simply throw here or use `BuildErrorResponse`

Reply:
I suspect that this will only work for outbound connection. The endpoint is a IP:Port and the default port will only be correct when considering outbound connections. For an inbound connection peer will connect using ephemeral  port number.